### PR TITLE
Integration testing and optimization

### PR DESCRIPTION
**Pull request by jaferIdrees**
### Project requirements
n+1

- [ ] Make sure that the N+1 problem is solved when fetching all posts and their comments for a user. This is the corresponding wireframeProject requirements

- Use Capybara to write integration tests for each view in your project.

> - User index page:

- [ ] I can see the username of all other users.
- [ ] I can see the profile picture for each user.
- [ ] I can see the number of posts each user has written.
- [ ] When I click on a user, I am redirected to that user's show page.

> -  user show page:
- [ ] I can see the user's profile picture.
- [ ] I can see the user's username.
- [ ] I can see the number of posts the user has written.
- [ ] I can see the user's bio.
- [ ] I can see the user's first 3 posts.
- [ ] I can see a button that lets me view all of a user's posts.
- [ ] When I click a user's post, it redirects me to that post's show page.
- [ ] When I click to see all posts, it redirects me to the user's post's index page.

> - User post index page:
- [ ] I can see the user's profile picture.
- [ ] I can see the user's username.
- [ ] I can see the number of posts the user has written.
- [ ] I can see a post's title.
- [ ] I can see some of the post's body.
- [ ] I can see the first comments on a post.
- [ ] I can see how many comments a post has.
- [ ] I can see how many likes a post has.
- [ ] I can see a section for pagination if there are more posts than fit on the view.
- [ ] When I click on a post, it redirects me to that post's show page.

> - Post show page:
- [ ] I can see the post's title.
- [ ] I can see who wrote the post.
- [ ] I can see how many comments it has.
- [ ] I can see how many likes it has.
- [ ] I can see the post body.
- [ ] I can see the username of each comment.
- [ ] I can see the comment each comment left.

### CODE REVIEW

File: spec/features/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      expect(page.current_path).to eql('/users/1/posts/new')
+    end
+    it 'Is redirecting to all user post' do
+      click_link 'See all posts'
+      expect(page.current_path).to eql('/users/1/posts')
+    end
+    it 'I can some of the posts ' do
+      expect(page).to have_content 'Lorem ipsum dolor sit amet,'
+    end
+  end
```
Comment (TedLivist) [inline spec/features/user_show_spec.rb:33]: - [ ] You are missing User show tests tht assert the following scenarios:

  - Tests that asserts you can see the user's profile picture.
  - Tests that asserts you can see the user's username.
  - Tests that asserts you can see the number of posts the user has written.

File: spec/features/post_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      expect(page).to have_css('.social-post span', text: 'Likes: 9', count: 1)
+    end
+    it 'When I click on a post, it redirects me to that posts show page.' do
+      click_link 'Post-1'
+      expect(page.current_path).to eql('/users/1/posts/1')
+    end
+    it 'Is displaying the user all posts.' do
+      expect(page).to have_css('.post-card', count: 5)
+    end
+  end
```
Comment (TedLivist) [inline spec/features/post_index_spec.rb:32]: - [ ] You aremissing Post index tests that assert the following. Kindly add them

  - Tests that asserts you can see the user's profile picture.
  - Tests that asserts you can see the user's username.
  - Tests that asserts you can see the number of posts the user has written.
  - Tests that asserts you can see a post's title.
  - Tests that asserts you can see some of the post's body.
  - Tests that asserts you can see the first comments on a post.

**Review by Mihndim2020 — APPROVED** [review]
# STATUS: APPROVED!!!  🎉 🎉 🎉 :champagne: :champagne: :champagne:
### Hi @raminka13, 

### Good job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Good commit messages
:heavy_check_mark:  Clean and Organized code
:heavy_check_mark: Descriptive Readme file

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 :champagne: :champagne: :champagne:

Congratulations-Memes-FEATURED

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

### N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear or you can get to me on slack  using this handle @mihndim.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/post_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      click_button 'Add Comment'
+      expect(page).to have_css('.social-post > span', text: '📝 Comments: 10', count: 1)
+      expect(page).to have_content('Jack: New Comment')
+      expect(page.current_path).to eql('/users/1/posts/1')
+    end
+    it 'Is redirecting to all user post' do
+      click_link 'Back to posts'
+      expect(page.current_path).to eql('/users/1/posts')
+    end
+  end
```
Comment (TedLivist) [inline spec/features/post_show_spec.rb:33]: - [ ] You are missing Post show tests that assert the following:

  * Tests that asserts you can see the post's title.
  * Tests that asserts you can see who wrote the post.
  * Tests that asserts you can see how many comments it has.
  * Tests that asserts you can see how many likes it has.
  * Tests that asserts you can see the post body.
  * Tests that asserts you can see the username of each commentor.
  * Tests that asserts you can see the comment each commentor left.

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @raminka13 @jaferIdrees,

Good job so far!
✔️ All tests are passing
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**